Nrf91 Documentation update for using devicetree partitions. (#28318)
nrfconnect/sdk-nrf: doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_dt_integration.rst

.. _devicetree_integration:

Devicetree integration
#############################

The Modem library, which runs on the application core, shares an area of RAM memory with the modem core.
During the initialization, the Modem library accepts the boundaries of this area of RAM and configures the communication with the modem core accordingly.

> However, it is the responsibility of the application to reserve that RAM during linking, so that this memory area is not used for other purposes and remain dedicated for use by the Modem library.

Reviewer: ```suggestion
However, it is the responsibility of the application to reserve that RAM during linking, so that this memory area is not used for other purposes and remains dedicated for use by the Modem library.
```

> The size and location of the shared memory can be defined in device tree by creating the following ram sections:

Reviewer 2: ```suggestion
The size and location of the shared memory can be defined in device tree by creating the following RAM sections:
```

* ``cpuapp_cpucell_ipc_shm_ctrl`` is the control area.
* ``cpuapp_cpucell_ipc_shm_heap`` is used for data from libmodem to modem.
* ``cpucell_cpuapp_ipc_shm_heap`` is used for data from modem to libmodem.

> * ``cpucell_cpuapp_ipc_shm_trace`` is optional and used for tracing.

Reviewer 2: ```suggestion
* ``cpucell_cpuapp_ipc_shm_trace`` is optional and used for modem tracing.
```

The sections need to be placed in a specific area depending on the device:

* nrf91: within the first 128 kB of ram.

> * nrf92: Within the last 64 kB of ram.

Reviewer 2: And, should that be RAM0 instead for nRF92?
```suggestion
* nRF91: within the first 128 kB of RAM.
* nRF92: Within the last 64 kB of RAM.
```

> * nrf92: Within the last 64 kB of ram.

Reviewer: ```suggestion
* nRF91 Series devices: Within the first 128 kB of RAM.
* nRF92 Series devices: Within the last 64 kB of RAM.
```
To use more official terms

The control section has a minimum size depending on the modem version:

* nrf91 LTE: 1256 bytes
* nrf91 DECT: 1832 bytes

> * nrf92 LTE: 4096 bytes

Reviewer 2: ```suggestion
* nRF91 LTE: 1256 bytes
* nRF91 DECT: 1832 bytes
* nRF92 LTE: 4096 bytes
```

> * nrf92 LTE: 4096 bytes

Reviewer: You could maybe use the modem firmware product names here, that is, mfw_91x1, mfw-nr+_nrf91x1 and so on

Reviewer 2: I'm not sure if the name for nRF92 MFW has been agreed yet, I will check.

Author: Should we list NTN as well then? What will be the firmware name for the nRF9251? What about the nRF9280?
Would be nice but a bit of extra effort to maintain a complete table here.

Reviewer 2: If we change this, for nRF91 the list should be `mfw_nrf9160`, `mfw_nrf91x1` and `mfw_nrf9151-ntn`.

Reviewer 2: The official name for nRF92 firmware is `ciot_nrf92` (yes, no `mfw` in the name on purpose).

Author: I made it into a table.

Author: I have removed the nRF92 reference, for now we are only supposed to add code for the nRF92, not documentation. We can add the nRF92 line in the table again later.

.. important::
   The heap implementation used for allocations on the TX region has an overhead of up to 128 bytes.

>    Adjust the size of the TX region accordingly, so that its size is 128 bytes larger than the largest allocation you expect to happen (longest AT command, largest payload passed to :c:func:`nrf_send`) in your application.

Reviewer 2: It would be clearer to use the RAM section name instead.

.. _devicetree_integration:

Devicetree integration
#############################

The Modem library, which runs on the application core, shares a region of RAM memory with the modem core.
During the initialization, the Modem library accepts the boundaries of this region of RAM and configures the communication with the modem core accordingly.

However, it is the responsibility of the application to reserve that RAM during linking, so that this memory area is not used for other purposes and remains dedicated for use by the Modem library.

The size and location of the shared memory region can be defined in device tree by creating the following RAM sections:

* ``cpuapp_cpucell_ipc_shm_ctrl`` is used for control messages.
* ``cpuapp_cpucell_ipc_shm_heap`` is used for data from libmodem to modem.
* ``cpucell_cpuapp_ipc_shm_heap`` is used for data from modem to libmodem.
* ``cpucell_cpuapp_ipc_shm_trace`` is optional and used for modem tracing.

The regions shared with modem need to be placed in a specific area depending on the device:

> * nRF91 Series devices: within the first 128 kB of RAM.

Reviewer: ```suggestion
* nRF91 Series devices: Within the first 128 kB of RAM.
```

> * nRF91 Series devices: Within the first 128 kB of RAM.

Author: Many samples decrease or increase the shared memory with modem to either increase the RAM for the App or allow for larger packets. Other samples add the trace partition to the shared memory.

I think this would be a good place to document the limitations of the shared memory. Our internal documentation is a bit cryptic. @mparpala can you confirm this is correct and complete? Do we want to add the actual addresses? Then we need to split it into nRF9251 and nRF9280.

.. _devicetree_integration:

Devicetree integration

> #############################

Reviewer 3: ```suggestion
Devicetree integration
######################
```

> The Modem library, which runs on the application core, shares a region of RAM memory with the modem core.

Reviewer 4: Could link to the nrfxlib Modem library - https://docs.nordicsemi.com/bundle/ncs-latest/page/nrfxlib/nrf_modem/README.html.

During the initialization, the Modem library accepts the boundaries of this region of RAM and configures the communication with the modem core accordingly.

> However, it is the responsibility of the application to reserve that RAM during linking, so that this memory area is not used for other purposes and remains dedicated for use by the Modem library.

Reviewer 4: ```suggestion
However, it is the responsibility of the application to reserve that RAM during linking so that this memory area is not used for other purposes and remains dedicated for use by the Modem library.
```

> The size and location of the shared memory region can be defined in device tree by creating the following RAM sections:

Reviewer 4: ```suggestion
The size and location of the shared memory region can be defined in the devicetree by creating the following RAM sections:
```

* ``cpuapp_cpucell_ipc_shm_ctrl`` is used for control messages.
* ``cpuapp_cpucell_ipc_shm_heap`` is used for data from libmodem to modem.
* ``cpucell_cpuapp_ipc_shm_heap`` is used for data from modem to libmodem.
* ``cpucell_cpuapp_ipc_shm_trace`` is optional and used for modem tracing.

> The regions shared with modem need to be placed in a specific area depending on the device:

Reviewer 4: ```suggestion
The regions shared with the modem need to be placed in a specific area depending on the device:
```

* nRF91 Series devices: Within the first 128 kB of RAM.
* nRF92 Series devices: Within the last 64 kB of Global Domain RAM0.

> The control region has a minimum size depending on the modem version:

Reviewer 2: ```suggestion
The control region has a minimum size depending on the modem firmware:
```

> .. list-table:: Firmware Sizes

Reviewer 2: I don't think a caption is mandatory. If it's present, it should be something else.
```suggestion
.. list-table:: Control region sizes
```

   :widths: 50 50
   :header-rows: 1

   * - Firmware

>      - control region size

Reviewer 2: ```suggestion
     - Control region size
```

.. _devicetree_integration:

Devicetree integration
######################

The :ref:`nrfxlib:nrf_modem`, which runs on the application core, shares a region of RAM memory with the modem core.
During the initialization, the Modem library accepts the boundaries of this region of RAM and configures the communication with the modem core accordingly.

However, it is the responsibility of the application to reserve that RAM during linking so that this memory area is not used for other purposes and remains dedicated for use by the Modem library.

The size and location of the shared memory can be defined in devicetree by creating the following RAM sections:

> * ``sram0_ns_modem``

Reviewer 3: ```suggestion
* ``sram0_ns_modem``:
```

  This section contains all the subsections needed for communicating with the modem.

>   It must be placed in a specific area of RAM depending on the device (see the :ref:`table below <modem_shmem_variations>`).

Reviewer 4: ```suggestion
  You must place it in a specific area of RAM depending on the device (see the :ref:`following table <modem_shmem_variations>`).
```

>   * ``cpuapp_cpucell_ipc_shm_ctrl``

Reviewer 3: ```suggestion
  * ``cpuapp_cpucell_ipc_shm_ctrl``:
```

    This section is used for control messages.

>     It has a minimum size depending on the modem firmware (see the :ref:`table below <modem_shmem_variations>`).

Reviewer 4: ```suggestion
    It has a minimum size depending on the modem firmware (see the :ref:`following table <modem_shmem_variations>`).
```

>   * ``cpuapp_cpucell_ipc_shm_heap``

Reviewer 3: ```suggestion
  * ``cpuapp_cpucell_ipc_shm_heap``:
```

>     This section is used for data sent from libmodem to the modem.

Reviewer: ```suggestion
    This section is used for data sent from the Modem library to the modem.
```

    The heap implementation used has an overhead of up to 128 bytes.
    Adjust the size of this region so it is 128 bytes larger than the largest allocation you expect to happen in your application (such as the longest AT command or the largest payload passed to :c:func:`nrf_send`).

>   * ``cpucell_cpuapp_ipc_shm_heap``

Reviewer 3: ```suggestion
  * ``cpucell_cpuapp_ipc_shm_heap``:
```

>     This section is used for data sent from the modem to libmodem.

Reviewer: ```suggestion
    This section is used for data sent from the modem to the Modem library.
```

>   * ``cpucell_cpuapp_ipc_shm_trace``

Reviewer 3: ```suggestion
  * ``cpucell_cpuapp_ipc_shm_trace``:
```

    This section is used for modem tracing.
    It is optional and only needs to be added when modem tracing is enabled.

> * ``sram0_ns_app``

Reviewer 3: ```suggestion
* ``sram0_ns_app``:
```
